heron/common/src/java/com/twitter/heron/common/basics/ByteAmount.java

```diff
@@ -0,0 +1,247 @@
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package com.twitter.heron.common.basics;
+
+/**
+ * Class that encapsulates number of bytes, with helpers to handle units properly.
+ */
+public final class ByteAmount implements Comparable<ByteAmount> {
+  private static final long MB = 1024L * 1024;
+  private static final long GB = MB * 1024;
+  private static final long MAX_MB = Math.round(Long.MAX_VALUE / MB);
+  private static final long MAX_GB = Math.round(Long.MAX_VALUE / GB);
+
+  public static final ByteAmount ZERO = ByteAmount.fromBytes(0);
+  private final long bytes;
+
+  private ByteAmount(long bytes) {
+    this.bytes = bytes;
+  }
+
+  /**
+   * Creates a ByteAmount value in bytes.
+   *
+   * @param bytes value in bytes to represent
+   * @return a ByteAmount object repressing the number of bytes passed
+   */
+  public static ByteAmount fromBytes(long bytes) {
+    return new ByteAmount(bytes);
+  }
+
+  /**
+   * Creates a ByteAmount value in megabytes. If the megabytes value
+   * is >= Long.MAX_VALUE / 1024 / 1024, the byte representation is capped at Long.MAX_VALUE.
+   *
+   * @param megabytes value in megabytes to represent
+   * @return a ByteAmount object repressing the number of MBs passed
+   */
+  public static ByteAmount fromMegabytes(long megabytes) {
+    if (megabytes >= MAX_MB) {
+      return new ByteAmount(Long.MAX_VALUE);
+    } else {
+      return new ByteAmount(megabytes * MB);
+    }
+  }
+
+  /**
+   * Creates a ByteAmount value in gigabytes. If the gigabytes value
+   * is >= Long.MAX_VALUE / 1024 / 1024 / 1024, the byte representation is capped at Long.MAX_VALUE.
+   *
+   * @param gigabytes value in gigabytes to represent
+   * @return a ByteAmount object repressing the number of GBs passed
+   */
+  public static ByteAmount fromGigabytes(long gigabytes) {
+    if (gigabytes >= MAX_GB) {
+      return new ByteAmount(Long.MAX_VALUE);
+    } else {
+      return new ByteAmount(gigabytes * GB);
+    }
+  }
+
+  /**
+   * Returns the number of bytes associated with the ByteValue.
+   * @return number of bytes
+   */
+  public long asBytes() {
+    return bytes;
+  }
+
+  /**
+   * Converts the number of bytes to megabytes, rounding if there is a remainder. Because of loss
+   * of precision due to rounding, it's strongly advised to only use this method when it is certain
+   * that megabytes were using during the construction of the object via fromMegabytes. Only In that
+   * scenario is there a guarantee that precision will not be lost when calling this method.
+   * @return returns the ByteValue in MBs or 0 if the value is < (1024 * 1024) / 2
+   */
+  public long asMegabytes() {
```

Comment:
I debated this one as well. I like the conciseness but not the resulting inconsistency between asBytes, asMB and asGB.